coderd/notifications/notifier.go

```diff
@@ -96,30 +109,79 @@ func (n *notifier) run(success chan<- dispatchResult, failure chan<- dispatchRes
 		n.log.Info(context.Background(), "gracefully stopped")
 	}()
 
-	// TODO: idea from Cian: instead of querying the database on a short interval, we could wait for pubsub notifications.
-	//		 if 100 notifications are enqueued, we shouldn't activate this routine for each one; so how to debounce these?
-	//		 PLUS we should also have an interval (but a longer one, maybe 1m) to account for retries (those will not get
-	//		 triggered by a code path, but rather by a timeout expiring which makes the message retryable)
-
-	// run the ticker with the graceful context, so we stop fetching after stop() is called
-	tick := n.clock.TickerFunc(n.gracefulCtx, n.cfg.FetchInterval.Value(), func() error {
-		// Check if notifier is not paused.
-		ok, err := n.ensureRunning(n.outerCtx)
-		if err != nil {
-			n.log.Warn(n.outerCtx, "failed to check notifier state", slog.Error(err))
-		}
-
-		if ok {
-			err = n.process(n.outerCtx, success, failure)
+	// loopTick is used to synchronize the goroutine that processes messages with the ticker.
+	loopTick := make(chan chan struct{})
+	// loopDone is used to signal when the processing loop has exited due to
+	// graceful stop or otherwise.
+	loopDone := make(chan struct{})
+	go func() {
+		defer close(loopDone)
+		for c := range loopTick {
```

Comment:
Might need to check if the ctx is done at the top of each iteration, or at least before the blocking call to `process()`